includes/wc-order-functions.php

```diff
@@ -689,17 +689,23 @@ function wc_restock_refunded_items( $order, $refunded_line_items ) {
 		if ( ! isset( $refunded_line_items[ $item_id ], $refunded_line_items[ $item_id ]['qty'] ) ) {
 			continue;
 		}
-		$product = $item->get_product();
+		$product            = $item->get_product();
+		$item_stock_reduced = $item->get_meta( '_reduced_stock', true );
 
-		if ( $product && $product->managing_stock() ) {
-			$old_stock = $product->get_stock_quantity();
-			$new_stock = wc_update_product_stock( $product, $refunded_line_items[ $item_id ]['qty'], 'increase' );
+		if ( ! $item_stock_reduced || ! $product || ! $product->managing_stock() ) {
+			continue;
+		}
 
-			/* translators: 1: product ID 2: old stock level 3: new stock level */
-			$order->add_order_note( sprintf( __( 'Item #%1$s stock increased from %2$s to %3$s.', 'woocommerce' ), $product->get_id(), $old_stock, $new_stock ) );
+		$old_stock = $product->get_stock_quantity();
+		$new_stock = wc_update_product_stock( $product, $refunded_line_items[ $item_id ]['qty'], 'increase' );
 
-			do_action( 'woocommerce_restock_refunded_item', $product->get_id(), $old_stock, $new_stock, $order, $product );
-		}
+		/* translators: 1: product ID 2: old stock level 3: new stock level */
+		$order->add_order_note( sprintf( __( 'Item #%1$s stock increased from %2$s to %3$s.', 'woocommerce' ), $product->get_id(), $old_stock, $new_stock ) );
+
+		$item->delete_meta_data( '_reduced_stock' );
```

Comment:
I think this does not cater for some scenarios:

### Partial refund
1. Order 4 products P ... `_reduced_stock` set to 4, product qty -4
2. Refund 2 products P ... `_reduced_stock` deleted, product qty +2
3. Cancel order ... no `_reduced_stock`, product qty +0

### Multi-refund
1. Order 4 products P ... `_reduced_stock` set to 4
2. Refund 1 products P ... `_reduced_stock` deleted, product qty +1
2. Refund 1 products P ... no `_reduced_stock`, product qty +0
3. Cancel order ... no `_reduced_stock`, product qty +0

Reply:
You're right. I will borrow the correct logic for https://github.com/woocommerce/woocommerce/pull/22329/files since I worked on that in a different place already.

Reply:
Sorted